correction-station-i2c/correction_station_i2c.go

```diff
@@ -0,0 +1,182 @@
+	I2CBaudRate int `json:"i2c_baud_rate,omitempty"`
+}
+
+// Validate ensures all parts of the config are valid.
+func (cfg *Config) Validate(path string) ([]string, error) {
+	var deps []string
+	if cfg.RequiredAccuracy == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "required_accuracy")
+	}
+	if cfg.RequiredAccuracy < 0 || cfg.RequiredAccuracy > 5 {
+		return nil, errRequiredAccuracy
+	}
+	if cfg.RequiredTime == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "required_time")
+	}
+
+	if cfg.I2CBus == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "i2c_bus")
+	}
+	if cfg.I2CAddr == 0 {
+		return nil, utils.NewConfigValidationFieldRequiredError(path, "i2c_addr")
+	}
+
+	return deps, nil
+}
+
+type rtkStationI2C struct {
+	resource.Named
+	resource.AlwaysRebuild
+	logger           golog.Logger
+	correctionSource correctionSource
+	protocol         string
+	i2cPath          i2cBusAddr
+
+	cancelCtx               context.Context
+	cancelFunc              func()
+	activeBackgroundWorkers sync.WaitGroup
+
+	err movementsensor.LastError
+}
+
+type correctionSource interface {
+	Start(ready chan<- bool)
+	Close(ctx context.Context) error
+}
+
+type i2cBusAddr struct {
+	bus  int
+	addr byte
+}
+
+func newRTKStationI2C(
+	ctx context.Context,
+	deps resource.Dependencies,
+	name resource.Name,
+	newConf *Config,
+	logger golog.Logger,
+) (sensor.Sensor, error) {
+
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+
+	r := &rtkStationI2C{
+		Named:      name.AsNamed(),
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+		logger:     logger,
+		err:        movementsensor.NewLastError(1, 1),
+	}
+
+	err := ConfigureBaseRTKStation(newConf)
+	if err != nil {
+		r.logger.Info("rtk base station could not be configured")
+		return r, err
+	}
+
+	// Init correction source
+	r.i2cPath.addr = byte(newConf.I2CAddr)
+	r.i2cPath.bus = newConf.I2CBus
+	r.correctionSource, err = newI2CCorrectionSource(deps, newConf, logger)
+	if err != nil {
+		return nil, err
+	}
+	r.logger.Debug("Starting")
+
+	r.start(ctx)
+	return r, r.err.Get()
+}
+
+// Start starts reading from the correction source and sends corrections the i2c buffer.
+func (r *rtkStationI2C) start(ctx context.Context) {
+	r.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer r.activeBackgroundWorkers.Done()
+
+		if err := r.cancelCtx.Err(); err != nil {
+			return
+		}
+
+		// read from correction source
+		ready := make(chan bool)
+		r.correctionSource.Start(ready)
+
+		select {
+		case <-ready:
+		case <-r.cancelCtx.Done():
+			return
+		}
+	})
+}
+
+// Close shuts down the rtkStation.
+func (r *rtkStationI2C) Close(ctx context.Context) error {
+	r.cancelFunc()
+	r.activeBackgroundWorkers.Wait()
+
+	// close correction source
+	err := r.correctionSource.Close(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := r.err.Get(); err != nil && !errors.Is(err, context.Canceled) {
```

Comment:
I wonder if checking context.Canceled is necessary here or we should check if context is canceled separately.

Reply:
If the error is context cancelled, you'll close with an error. You can be expecting a context cancellation from the for loop upon closing since that's how contact cancels itself. We can check or not check as long as it closes.

If we check it elsewhere, this err being context cancelled and not nil will return that upon closing.